### ISSUE TITLE
Function body is ignored by `mypy`

### ISSUE THREAD

**Issue by fjarri** [python/mypy#14484]
**Bug Report**

Under certain circumstances bodies of some functions are ignored during the typecheck without any warning messages.

**To Reproduce**

(`t.py`)
```python
import numpy

from typing import List, Any, Protocol, runtime_checkable

@runtime_checkable
class P(Protocol):
    dtype: List[Any]

def foo(arr: "numpy.ndarray[Any, Any]") -> None:
    assert isinstance(arr, P)
    reveal_type(arr)
    1 + "A"
```
```
$ mypy t.py
```
(same behavior with `mypy --strict`)

**Expected Behavior**
- `reveal_type` message in the output
- typing error for `1 + "A"`

The expected behavior can be observed if one changes the MRE slightly, e.g. replaces `dtype: List[Any]` with `dtype: Any`:
```
t.py:13: note: Revealed type is "t.<subclass of "ndarray" and "P">"
t.py:14: error: Unsupported operand types for + ("int" and "str")  [operator]
Found 1 error in 1 file (checked 1 source file)
```

**Actual Behavior**
```
Success: no issues found in 1 source file
```

**Your Environment**
Python 3.8.13
mypy 0.991
numpy 1.24.1

**Comment by hauntsaninja** [comment]
This is due to https://mypy.readthedocs.io/en/stable/common_issues.html#unreachable-code

Maybe use the `--warn-unreachable` flag

**Comment by fjarri** [comment]
The problem is that these lines are reachable in the original code (that's how I noticed it - the tests were failing). E.g. this slightly more complicated example
```python
import numpy

from typing import List, Any, Protocol, runtime_checkable

@runtime_checkable
class P(Protocol):
    dtype: "numpy.dtype[Any]"

def foo(arr: "numpy.ndarray[Any, Any]") -> None:
    assert isinstance(arr, P)
    reveal_type(arr)
    1 + "A"

foo(numpy.ones(10))
```
when executed, reaches the `reveal_type()` statement.

This does not reproduce with a manually constructed type with a single field and a protocol containing that field, so I'm assuming something in `numpy` throws `mypy` off?

**Comment by hauntsaninja** [comment]
Yeah, the fact that it's reachable at runtime doesn't mean too much, since runtime_checkable Protocols provide pretty weak guarantees.

It looks like the issue is that ndarray.dtype is typed as a property: https://github.com/numpy/numpy/blob/df3751b03d789ee04cac3a9a8a7f7f3aba58735c/numpy/__init__.pyi#L2522

So your protocol typing it as an attribute means it's not compatible, because e.g. you can't assign to it.

Instead, try the following:
```
class P(Protocol):
    @property
    def dtype(self) -> "numpy.dtype[Any]": ...
```

**Comment by fjarri** [comment]
Technically you can assign to array's `dtype`, but I don't need it in my code, so if the variant with `@property` works, that's even better. Thank you for your help! I am surprised that `--strict` does not include the reachability check already, but I'll just enable it in the config.
